(maint) Simplify Travis scripts and add check for commit summaries

MikaelSmith opened the pull request:
Separates Doxygen generation as its own step, and simplifies the script
based on puppetlabs/cpp-project-template. Only installs files that are
necessary for the particular build task being performed.

peterhuene commented on scripts/travis_target.sh:
```diff
@@ -1,85 +1,75 @@
     fi
+  done
+  exit 0
+elif [ ${TRAVIS_TARGET} == DEBUG ]; then
```
Shouldn't this be RELEASE?  The release build in Travis built debug with these changes.

MikaelSmith replied:
Oh right, oops.

MikaelSmith replied:
Fixed by swapping cmake invocations.

MikaelSmith commented:
Rebased for the doxygen changes, and made some fixes to the regex (it used to still check for pup-, \d doesn't appear to be supported in Bash regex) and added "Revert".

MikaelSmith commented on scripts/travis_target.sh:
```diff
@@ -1,84 +1,72 @@
-}
+  if [ ${TRAVIS_TARGET} == DEBUG ]; then
+    # Ignore coveralls failures, keep service success uncoupled
+    coveralls --gcov gcov-4.8 --gcov-options '\-lp' -e vendor >/dev/null || true
```
Moving the build dir to root includes files that aren't compiled on this platform. It's a more complete view of what's tested in Travis, but doesn't include testing from Windows.

MikaelSmith commented:
Updated with fix to commit title regex based on https://github.com/puppetlabs/hiera/pull/284.
